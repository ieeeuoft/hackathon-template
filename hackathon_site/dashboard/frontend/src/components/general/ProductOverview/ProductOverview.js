import React from "react";
import Typography from "@material-ui/core/Typography";
import Link from "@material-ui/core/Link";
import Button from "@material-ui/core/Button";
import MenuItem from "@material-ui/core/MenuItem";
import FormControl from "@material-ui/core/FormControl";
import Select from "@material-ui/core/Select";
import LaunchIcon from "@material-ui/icons/Launch";
import InputLabel from "@material-ui/core/InputLabel";
import Chip from "@material-ui/core/Chip";
import styles from "./ProductOverview.module.scss";
import Alert from "@material-ui/lab/Alert";

import { Formik } from "formik";

const createQuantityList = (number) => {
    let entry = [];

    for (let i = 1; i <= number; i++) {
        entry.push(
            <MenuItem key={i} role="quantity" value={i.toString()}>
                {i}
            </MenuItem>
        );
    }

    return entry;
};

export const CartForm = ({
    availableQuantity,
    handleSubmit,
    handleChange,
    requestFailure,
    values: { quantity },
}) => {
    return (
        <>
            {requestFailure && (
                <Alert className={styles.alert} variant="filled" severity="error">
                    {requestFailure.message}
                </Alert>
            )}
            <form className={styles.form} onSubmit={handleSubmit}>
                <FormControl variant="outlined" className={styles.formControl}>
                    <InputLabel>Qty</InputLabel>
                    <Select
                        value={quantity}
                        role="selecter"
                        onChange={handleChange}
                        label="Qty"
                        name="quantity"
                    >
                        {createQuantityList(availableQuantity)}
                    </Select>
                </FormControl>
                <Button
                    variant="contained"
                    className={styles.cartButton}
                    type="submit"
                    onClick={handleSubmit}
                    disableElevation
                >
                    ADD TO CART
                </Button>
            </form>
        </>
    );
};

export const EnhancedCartForm = ({
    handleSubmit,
    requestFailure,
    availableQuantity,
}) => {
    const onSubmit = (formikValues) => {
        handleSubmit(formikValues.quantity);
    };

    return (
        handleSubmit && (
            <Formik
                initialValues={{ quantity: "1" }}
                onSubmit={onSubmit}
                validateOnBlur={false}
                validationOnChange={false}
                validationSchema={false}
            >
                {(formikProps) => (
                    <CartForm
                        availableQuantity={availableQuantity}
                        handleSubmit={formikProps.handleSubmit}
                        handleChange={formikProps.handleChange}
                        requestFailure={requestFailure}
                        values={formikProps.values}
                    />
                )}
            </Formik>
        )
    );
};

const DetailInfoSection = ({
    manufacturer,
    model_num,
    datasheet,
    notes,
    constraints,
}) => {
    return (
        <div className={styles.detailinfosection}>
            <div className={styles.bodyinfo}>
                <Typography variant="body2" className={styles.heading}>
                    Constraints
                </Typography>
                {constraints.map((constraint, i) => (
                    <Typography key={i} variant="body2">
                        {constraint}
                    </Typography>
                ))}
            </div>
            <div className={styles.bodyinfo}>
                <Typography variant="body2" className={styles.heading}>
                    Manufacturer
                </Typography>
                <Typography>{manufacturer}</Typography>
            </div>
            <div className={styles.bodyinfo}>
                <Typography variant="body2" className={styles.heading}>
                    Model Number
                </Typography>
                <Typography>{model_num}</Typography>
            </div>
            <div className={styles.bodyinfo}>
                <Typography variant="body2" className={styles.heading}>
                    Datasheet
                </Typography>
                <Link
                    href={datasheet}
                    rel="noopener"
                    color="inherit"
                    underline="none"
                    target="_blank"
                    className={styles.bodyinfoDataSheet}
                >
                    <LaunchIcon></LaunchIcon>
                    <Typography>Link</Typography>
                </Link>
            </div>
            <div className={styles.bodyinfo}>
                <Typography variant="body2" className={styles.heading}>
                    Notes
                </Typography>
                {notes.map((note, i) => (
                    <Typography key={i} variant="body2">
                        {note}
                    </Typography>
                ))}
            </div>
        </div>
    );
};

const MainSection = ({ type, name, total, available, categories, img }) => {
    let availability;
    if (available === 0) {
        availability = (
            <Typography className={styles.notAvailable}>OUT OF STOCK</Typography>
        );
    } else {
        availability = (
            <Typography className={styles.available}>
                {available} OF {total} IN STOCK
            </Typography>
        );
    }

    return (
        <div className={styles.mainsection}>
            <div>
                <div className={styles.title}>
                    <Typography variant="h2">{name}</Typography>
                </div>
                {availability}
                <div className={styles.category}>
                    <Typography variant="h2">Category</Typography>
                    {categories.map((category, i) => (
                        <Chip
                            label={category}
                            size="small"
                            className={styles.categoryItem}
                            key={i}
                        />
                    ))}
                </div>
            </div>
            <img src={img} alt="product" />
        </div>
    );
};

export const ProductOverview = ({ detail, addToCart }) => {
    return (
        <div>
            <MainSection
                type={detail.type}
                name={detail.name}
                total={detail.total}
                available={detail.available}
                categories={detail.category}
                img={detail.img}
            />
            <DetailInfoSection
                manufacturer={detail.manufacturer}
                model_num={detail.model_num}
                datasheet={detail.datasheet}
                notes={detail.notes}
                constraints={detail.constraints}
            />
            <CartForm
                handleSubmit={addToCart}
                requestFailure={false}
                values={{}}
                availableQuantity={3}
            />
        </div>
    );
};
