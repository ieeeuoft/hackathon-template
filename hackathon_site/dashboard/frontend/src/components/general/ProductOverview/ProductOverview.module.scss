@import "./../../../assets/abstracts/mixins";
@import "./../../../assets/abstracts/variables";

.headersheet {
    display: flex;
    margin: 108px 50px 0 50px;
    .title {
        display: flex;
    }
    .available {
        margin: 9px 151px 0 0;
    }
    .category {
        margin-top: 59px;
        &Item {
            color: color(primary);
            background-color: color(blueLight);
            margin: 8px 2.5px;
        }
    }
    & img {
        margin: 0 11px;
    }
}

.bodysheet {
    margin: 10px 50px 0 50px;
    .bodyinfo {
        display: flex;
        flex-direction: column;
        h3 {
            margin: 5px 0;
        }
        margin-bottom: 20px;
        &DataSheet {
            display: flex;
        }
    }
}

.form {
    display: flex;
    flex-direction: column;
    &Control {
        margin: 20px 50px;
    }
    .cartButton {
        color: color(white);
        background-color: color(primary);
        margin: 30px 50px;
        padding: 10px 25px;
        &:hover {
            background-color: color(dark1);
        }
    }
}
