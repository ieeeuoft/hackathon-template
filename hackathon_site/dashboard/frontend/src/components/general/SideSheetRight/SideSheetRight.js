import React from "react";
import Drawer from "@material-ui/core/Drawer";
import Typography from "@material-ui/core/Typography";
import IconButton from "@material-ui/core/IconButton";
import ArrowBackIcon from "@material-ui/icons/ArrowBack";
import styles from "./SideSheetRight.module.scss";

const SideSheetRight = ({ children, title, isVisible, handleClose }) => {
    return (
        <Drawer anchor="right" open={isVisible} onClose={handleClose}>
            <div className={styles.header}>
                <IconButton
                    className={styles.headerBackArrow}
                    onClick={handleClose}
                    role="close"
                >
                    <ArrowBackIcon />
                </IconButton>
                <Typography variant="h2" role="title">
                    {title}
                </Typography>
            </div>
            <div className={styles.body}>{children}</div>
        </Drawer>
    );
};

export default SideSheetRight;
