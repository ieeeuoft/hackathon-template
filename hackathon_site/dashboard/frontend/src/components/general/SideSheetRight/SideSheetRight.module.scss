@import "./../../../assets/abstracts/mixins";
@import "./../../../assets/abstracts/variables";

.header {
    @include flexPosition($alIte: center);
    width: 100%;
    background-color: color(light2);
    z-index: 5;
    padding: 8px 34px;

    @include responsive(sm-down) {
        padding: 4px 12px;
    }

    &BackArrow {
        margin-right: 5px;
    }

    > h2 {
        @include responsive(md-down) {
            font-size: size(md) !important;
        }
    }
}

.body {
    width: 100%;
    height: 100%;
    padding: 25px 30px;

    @include responsive(sm-up) {
        padding: 45px 50px;
        min-width: 550px;
    }
}
