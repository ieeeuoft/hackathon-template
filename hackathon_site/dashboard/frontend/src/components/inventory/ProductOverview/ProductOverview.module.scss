@import "./../../../assets/abstracts/mixins";
@import "./../../../assets/abstracts/variables";

.heading {
    margin-top: 20px;
    font-weight: 500;
}

.productOverview {
    height: 100%;
    width: 100%;
    @include flexPosition($jusCon: space-between, $alIte: flex-start, $dir: column);

    &Div {
        width: 100%;
    }
}

.mainSection {
    @include flexPosition($jusCon: space-between);
    width: 100%;

    h6 {
        font-weight: 400;
    }

    .quantityAvailable {
        color: color(grey);
    }

    .categoryItem {
        background-color: color(blueLight);
        margin: 2.5px 5px 2.5px 0;
    }
    img {
        margin-left: 50px;
        width: 40%;
        max-width: 200px;
    }
}

.form {
    width: 100%;

    &Control {
        margin: 20px 0;
        width: 120px;
    }

    &Button {
        background-color: color(light2);
        width: calc(100% + 100px);
        margin-left: -50px;
        margin-bottom: -45px;
        padding: 30px 50px;

        @include responsive(sm-down) {
            width: calc(100% + 60px);
            margin-left: -30px;
            margin-bottom: -25px;
            padding: 20px 30px;
        }
    }
}
