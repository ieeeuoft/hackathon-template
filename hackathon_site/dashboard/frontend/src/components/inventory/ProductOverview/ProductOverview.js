import React from "react";
import Typography from "@material-ui/core/Typography";
import Button from "@material-ui/core/Button";
import MenuItem from "@material-ui/core/MenuItem";
import FormControl from "@material-ui/core/FormControl";
import Select from "@material-ui/core/Select";
import LaunchIcon from "@material-ui/icons/Launch";
import InputLabel from "@material-ui/core/InputLabel";
import Chip from "@material-ui/core/Chip";
import Alert from "@material-ui/lab/Alert";
import SideSheetRight from "components/general/SideSheetRight/SideSheetRight";

import * as Yup from "yup";
import { Formik } from "formik";

import styles from "./ProductOverview.module.scss";

export const ERROR_MESSAGES = {
    quantityMissing: "Quantity is required",
};

const addToCartFormSchema = Yup.object().shape({
    quantity: Yup.number().required(ERROR_MESSAGES.quantityMissing),
});

const createQuantityList = (number) => {
    let entry = [];

    for (let i = 1; i <= number; i++) {
        entry.push(
            <MenuItem key={i} role="quantity" value={i.toString()}>
                {i}
            </MenuItem>
        );
    }

    return entry;
};

export const AddToCartForm = ({
    quantityAvailable,
    constraintMax,
    handleSubmit,
    handleChange,
    requestFailure,
    values: { quantity },
}) => {
    const dropdownNum = !constraintMax
        ? quantityAvailable
        : Math.min(quantityAvailable, constraintMax);

    return (
        <>
            {requestFailure && (
                <Alert variant="filled" severity="error">
                    {requestFailure.message}
                </Alert>
            )}
            <form className={styles.form} onSubmit={handleSubmit}>
                <FormControl variant="outlined" className={styles.formControl}>
                    <InputLabel id="qtyLabel">Qty</InputLabel>
                    <Select
                        value={dropdownNum === 0 ? "" : quantity}
                        onChange={handleChange}
                        label="Qty"
                        labelId="qtyLabel"
                        name="quantity"
                        disabled={dropdownNum === 0}
                    >
                        {createQuantityList(dropdownNum)}
                    </Select>
                </FormControl>
                <div className={styles.formButton}>
                    <Button
                        variant="contained"
                        color="primary"
                        fullWidth={true}
                        size="large"
                        type="submit"
                        onClick={handleSubmit}
                        disabled={dropdownNum === 0}
                        disableElevation
                    >
                        Add to cart
                    </Button>
                </div>
            </form>
        </>
    );
};

export const EnhancedAddToCartForm = ({
    handleSubmit,
    requestFailure,
    quantityAvailable,
    constraintMax,
}) => {
    const onSubmit = (formikValues) => {
        handleSubmit(formikValues.quantity);
    };

    return (
        <Formik
            initialValues={{ quantity: "1" }}
            onSubmit={onSubmit}
            validateOnBlur={false}
            validationOnChange={false}
            validationSchema={addToCartFormSchema}
        >
            {(formikProps) => (
                <AddToCartForm
                    quantityAvailable={quantityAvailable}
                    constraintMax={constraintMax}
                    handleSubmit={formikProps.handleSubmit}
                    handleChange={formikProps.handleChange}
                    requestFailure={requestFailure}
                    values={formikProps.values}
                />
            )}
        </Formik>
    );
};

const DetailInfoSection = ({
    manufacturer,
    model_num,
    datasheet,
    notes,
    constraints,
}) => {
    return (
        <>
            <Typography variant="body2" color="secondary" className={styles.heading}>
                Constraints
            </Typography>
            {constraints.map((constraint, i) => (
                <Typography key={i}>{constraint}</Typography>
            ))}
            <Typography variant="body2" className={styles.heading}>
                Manufacturer
            </Typography>
            <Typography>{manufacturer}</Typography>
            <Typography variant="body2" className={styles.heading}>
                Model Number
            </Typography>
            <Typography>{model_num}</Typography>
            <Typography variant="body2" className={styles.heading}>
                Datasheet
            </Typography>
            <Button
                href={datasheet}
                rel="noopener"
                target="_blank"
                startIcon={<LaunchIcon />}
            >
                Link
            </Button>
            <Typography variant="body2" className={styles.heading}>
                Notes
            </Typography>
            {notes.split("\n").map((note, i) => (
                <Typography key={i}>{note}</Typography>
            ))}
        </>
    );
};

const MainSection = ({ name, total, quantityAvailable, categories, img }) => {
    const availability =
        quantityAvailable === 0 ? (
            <Typography color="secondary">OUT OF STOCK</Typography>
        ) : (
            <Typography className={styles.quantityAvailable}>
                {quantityAvailable} OF {total} IN STOCK
            </Typography>
        );

    return (
        <div className={styles.mainSection}>
            <div>
                <Typography variant="h6">{name}</Typography>
                {availability}
                <Typography variant="body2" className={styles.heading}>
                    Category
                </Typography>
                <div>
                    {categories.map((category, i) => (
                        <Chip
                            label={category}
                            size="small"
                            className={styles.categoryItem}
                            key={i}
                        />
                    ))}
                </div>
            </div>
            <img src={img} alt={name} />
        </div>
    );
};

export const ProductOverview = ({ detail, addToCart, isVisible, handleClose }) => (
    <SideSheetRight
        title="Product Overview"
        isVisible={isVisible}
        handleClose={handleClose}
    >
        <div className={styles.productOverview}>
            <div className={styles.productOverviewDiv}>
                <MainSection
                    type={detail.type}
                    name={detail.name}
                    total={detail.total}
                    quantityAvailable={detail.quantityAvailable}
                    categories={detail.category}
                    img={detail.img}
                />
                <DetailInfoSection
                    manufacturer={detail.manufacturer}
                    model_num={detail.model_num}
                    datasheet={detail.datasheet}
                    notes={detail.notes}
                    constraints={detail.constraints}
                />
            </div>

            {addToCart && (
                <EnhancedAddToCartForm
                    handleSubmit={addToCart}
                    requestFailure={false}
                    quantityAvailable={detail.quantityAvailable}
                    constraintMax={detail.constraintMax}
                />
            )}
        </div>
    </SideSheetRight>
);

export default ProductOverview;
