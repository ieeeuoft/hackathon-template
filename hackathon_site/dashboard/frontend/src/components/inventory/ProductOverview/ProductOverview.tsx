import React from "react";
import Typography from "@material-ui/core/Typography";
import Button from "@material-ui/core/Button";
import MenuItem from "@material-ui/core/MenuItem";
import FormControl from "@material-ui/core/FormControl";
import Select from "@material-ui/core/Select";
import LaunchIcon from "@material-ui/icons/Launch";
import InputLabel from "@material-ui/core/InputLabel";
import Chip from "@material-ui/core/Chip";
import CircularProgress from "@material-ui/core/CircularProgress";
import SideSheetRight from "components/general/SideSheetRight/SideSheetRight";

import * as Yup from "yup";
import { Formik, FormikValues } from "formik";

import styles from "./ProductOverview.module.scss";
import {
    closeProductOverview,
    displaySnackbar,
    isProductOverviewVisibleSelector,
} from "slices/ui/uiSlice";
import { useDispatch, useSelector } from "react-redux";
import { selectCategoriesByIds } from "slices/hardware/categorySlice";
import { RootState } from "slices/store";
import { addToCart, cartSelectors } from "slices/hardware/cartSlice";
import {
    hardwareInProductOverviewSelector,
    isUpdateDetailsLoading,
    removeProductOverviewItem,
} from "slices/hardware/hardwareSlice";
import { Category } from "api/types";
import hardwareImagePlaceholder from "assets/images/placeholders/no-hardware-image.svg";
import { hardwareSignOutEndDate, hardwareSignOutStartDate } from "constants.js";
import { Tooltip } from "@material-ui/core";
import { userTypeSelector, isTestUserSelector } from "slices/users/userSlice";

export const ERROR_MESSAGES = {
    quantityMissing: "Quantity is required",
};

const addToCartFormSchema = Yup.object().shape({
    quantity: Yup.number().required(ERROR_MESSAGES.quantityMissing),
});

const createQuantityList = (number: number) => {
    let entry = [];

    for (let i = 1; i <= number; i++) {
        entry.push(
            <MenuItem key={i} role="quantity" value={i.toString()}>
                {i}
            </MenuItem>
        );
    }

    return entry;
};

interface AddToCartFormProps extends FormikValues {
    quantityRemaining: number;
    maxPerTeam: number | null;
}
export const AddToCartForm = ({
    quantityRemaining,
    maxPerTeam,
    handleSubmit,
    handleChange,
    values: { quantity },
}: AddToCartFormProps) => {
    const isTestUser = useSelector(isTestUserSelector);
    const dropdownNum =
        maxPerTeam !== null
            ? Math.min(quantityRemaining, maxPerTeam)
            : quantityRemaining;

    const currentDateTime = new Date();
    const isOutsideSignOutPeriod =
        currentDateTime < hardwareSignOutStartDate ||
        currentDateTime > hardwareSignOutEndDate;

    const addToCartButton = (
        <Button
            variant="contained"
            color="primary"
            fullWidth={true}
            size="large"
            type="submit"
            onClick={handleSubmit}
            disabled={dropdownNum === 0 || (!isTestUser && isOutsideSignOutPeriod)}
            disableElevation
        >
            Add to cart
        </Button>
    );
    return (
        <>
            <form className={styles.form} onSubmit={handleSubmit}>
                <FormControl variant="outlined" className={styles.formControl}>
                    <InputLabel id="qtyLabel">Qty</InputLabel>
                    <Select
                        value={dropdownNum === 0 ? "" : quantity}
                        onChange={handleChange}
                        label="Qty"
                        labelId="qtyLabel"
                        name="quantity"
                        disabled={dropdownNum === 0}
                    >
                        {createQuantityList(dropdownNum)}
                    </Select>
                </FormControl>
                <div className={styles.formButton}>
                    {isOutsideSignOutPeriod ? (
                        <Tooltip title="You cannot add items to cart because hardware sign out period has not begun or is already over">
                            <span> {addToCartButton} </span>
                        </Tooltip>
                    ) : (
                        addToCartButton
                    )}
                </div>
            </form>
        </>
    );
};

interface EnhancedAddToCartFormProps {
    quantityRemaining: number;
    hardwareId: number;
    name: string;
    maxPerTeam: number | null;
}
export const EnhancedAddToCartForm = ({
    quantityRemaining,
    hardwareId,
    name,
    maxPerTeam,
}: EnhancedAddToCartFormProps) => {
    const dispatch = useDispatch();
    const currentQuantityInCart =
        useSelector((state: RootState) => cartSelectors.selectById(state, hardwareId))
            ?.quantity ?? 0;

    const onSubmit = (formikValues: { quantity: string }) => {
        const numQuantity: number = parseInt(formikValues.quantity);
        if (currentQuantityInCart + numQuantity <= (maxPerTeam ?? quantityRemaining)) {
            dispatch(addToCart({ hardware_id: hardwareId, quantity: numQuantity }));
            dispatch(
                displaySnackbar({
                    message: `Added ${numQuantity} ${name} item(s) to your cart.`,
                    options: {
                        variant: "success",
                    },
                })
            );
        } else {
            dispatch(
                displaySnackbar({
                    message: `Adding this amount to your cart will exceed the quantity limit for this item.`,
                    options: {
                        variant: "warning",
                    },
                })
            );
        }
    };

    return (
        <Formik
            initialValues={{ quantity: "1" }}
            onSubmit={onSubmit}
            validateOnBlur={false}
            validationOnChange={false}
            validationSchema={addToCartFormSchema}
        >
            {(formikProps) => (
                <>
                    {currentQuantityInCart > 0 && (
                        <Typography
                            variant="body2"
                            color="primary"
                            className={styles.heading}
                        >
                            You currently have {currentQuantityInCart} of this item in
                            your cart.
                        </Typography>
                    )}
                    <AddToCartForm
                        quantityRemaining={quantityRemaining}
                        maxPerTeam={maxPerTeam}
                        handleSubmit={formikProps.handleSubmit}
                        handleChange={formikProps.handleChange}
                        values={formikProps.values}
                    />
                </>
            )}
        </Formik>
    );
};

interface DetailInfoSectionProps {
    manufacturer: string;
    modelNumber: string;
    datasheet: string;
    notes?: string;
    constraints: string[];
}
export const DetailInfoSection = ({
    manufacturer,
    modelNumber,
    datasheet,
    notes,
    constraints,
}: DetailInfoSectionProps) => {
    return (
        <>
            {constraints?.length > 0 && (
                <>
                    <Typography
                        variant="body2"
                        color="secondary"
                        className={styles.heading}
                    >
                        Constraints
                    </Typography>
                    {constraints.map((constraint, i) => (
                        <Typography key={i}>- {constraint}</Typography>
                    ))}
                </>
            )}
            <Typography variant="body2" className={styles.heading}>
                Manufacturer
            </Typography>
            <Typography>{manufacturer}</Typography>
            {modelNumber && (
                <>
                    <Typography variant="body2" className={styles.heading}>
                        Model Number
                    </Typography>
                    <Typography>{modelNumber}</Typography>
                </>
            )}
            {datasheet && (
                <>
                    <Typography variant="body2" className={styles.heading}>
                        Datasheet
                    </Typography>
                    <Button
                        href={datasheet}
                        rel="noopener"
                        target="_blank"
                        startIcon={<LaunchIcon />}
                    >
                        Link
                    </Button>
                </>
            )}
            {notes && (
                <>
                    <Typography variant="body2" className={styles.heading}>
                        Notes
                    </Typography>
                    {notes.split("\n").map((note, i) => (
                        <Typography key={i}>{note}</Typography>
                    ))}
                </>
            )}
        </>
    );
};

interface MainSectionProps {
    name: string;
    quantityAvailable: number;
    quantityRemaining: number;
    categories: string[];
    picture: string;
}
const MainSection = ({
    name,
    quantityAvailable,
    quantityRemaining,
    categories,
    picture,
}: MainSectionProps) => {
    const userType = useSelector(userTypeSelector);
    const availability =
        quantityRemaining === 0 ? (
            <Typography color="secondary">OUT OF STOCK</Typography>
        ) : userType === "participant" ? (
            <Typography className={styles.quantityAvailable}>
                {Math.max(quantityRemaining, 0)} IN STOCK
            </Typography>
        ) : (
            <Typography className={styles.quantityAvailable}>
                {Math.max(quantityRemaining, 0)} OF {quantityAvailable} IN STOCK
            </Typography>
        );

    return (
        <div className={styles.mainSection}>
            <div>
                <Typography variant="h6">{name}</Typography>
                {availability}
                {categories.length > 0 && (
                    <>
                        <Typography variant="body2" className={styles.heading}>
                            Category
                        </Typography>
                        <div>
                            {categories.map((category, i) => (
                                <Chip
                                    label={category}
                                    size="small"
                                    className={styles.categoryItem}
                                    key={i}
                                />
                            ))}
                        </div>
                    </>
                )}
            </div>
            <img src={picture} alt={name} />
        </div>
    );
};

export const ProductOverview = ({
    showAddToCartButton,
}: {
    showAddToCartButton: boolean;
}) => {
    let categoryNames: string[] = [];
    let maxPerTeam: number | null = null;
    let constraints: string[] = [];

    const isLoading = useSelector(isUpdateDetailsLoading);

    const dispatch = useDispatch();
    const closeProductOverviewPanel = () => {
        dispatch(closeProductOverview());
        dispatch(removeProductOverviewItem());
    };

    const isProductOverviewVisible: boolean = useSelector(
        isProductOverviewVisibleSelector
    );
    const hardware = useSelector(hardwareInProductOverviewSelector);
    const categories = useSelector((state: RootState) =>
        selectCategoriesByIds(state, hardware?.categories || [])
    );

    maxPerTeam = hardware?.max_per_team ?? null;
    constraints = !!hardware?.max_per_team
        ? [`Max ${hardware.max_per_team} of this item`]
        : [];

    if (categories.length > 0) {
        categoryNames = categories
            .filter((category): category is Category => !!category)
            .map((category) => category.name);
        for (const category of categories) {
            if (category?.max_per_team !== undefined) {
                constraints.push(
                    `Max ${category.max_per_team} of items under category ${category.name}`
                );
                maxPerTeam =
                    maxPerTeam === null
                        ? category.max_per_team
                        : Math.min(category.max_per_team, maxPerTeam);
            }
        }
    }

    return (
        <SideSheetRight
            title="Product Overview"
            isVisible={isProductOverviewVisible}
            handleClose={closeProductOverviewPanel}
        >
            {isLoading ? (
                <CircularProgress
                    data-testid="circular-progress"
                    style={{
                        position: "absolute",
                        left: "50%",
                        top: "50%",
                    }}
                />
            ) : hardware ? (
                <div className={styles.productOverview}>
                    <div className={styles.productOverviewDiv}>
                        <MainSection
                            name={hardware.name}
                            quantityAvailable={hardware.quantity_available}
                            quantityRemaining={hardware.quantity_remaining}
                            categories={categoryNames}
                            picture={
                                hardware.picture ??
                                hardware.image_url ??
                                hardwareImagePlaceholder
                            }
                        />
                        <DetailInfoSection
                            manufacturer={hardware.manufacturer}
                            modelNumber={hardware.model_number}
                            datasheet={hardware.datasheet}
                            notes={hardware.notes}
                            constraints={constraints}
                        />
                    </div>

                    {showAddToCartButton && (
                        <EnhancedAddToCartForm
                            quantityRemaining={hardware.quantity_remaining}
                            hardwareId={hardware.id}
                            name={hardware.name}
                            maxPerTeam={maxPerTeam}
                        />
                    )}
                </div>
            ) : (
                <Typography variant="subtitle2" align="center" paragraph>
                    Unable to display hardware. Please refresh the page and try again.
                </Typography>
            )}
        </SideSheetRight>
    );
};

export default ProductOverview;
