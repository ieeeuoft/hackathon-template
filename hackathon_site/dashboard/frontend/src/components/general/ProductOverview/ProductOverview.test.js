import React from "react";
import SideSheetRight from "../SideSheetRight/SideSheetRight";
import { ProductOverview, EnhancedCartForm } from "./ProductOverview";
import {
    render,
    fireEvent,
    waitForElementToBeRemoved,
    waitFor,
    waitForElement,
} from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";

describe("Product Overview", () => {
    const testDetail = {
        name: "Arduino",
        type: "red",
        total: 30,
        available: 29,
        img: "https://www.filterforge.com/more/help/images/size200.jpg",
        category: ["MCU", "FPGA"],
        manufacturer: "Canakit",
        model_num: "Model 3B+",
        datasheet: "link",
        notes: ["- For micropython ask for image", "- randomnerdtutorials.com"],
        constraints: ["- Max 1 of this item", "- Max 3 microcontroller labelled red"],
        quantity: 3,
    };

    test("render", () => {
        const addCartMock = jest.fn();

        const { asFragment } = render(
            <SideSheetRight title="Product Overview" isVisible={true}>
                <ProductOverview detail={testDetail} addCartFunction={addCartMock} />
            </SideSheetRight>
        );

        expect(asFragment()).toMatchSnapshot();
    });

    test("add to cart button calls the correct function", () => {
        const addCartMock = jest.fn();

        const { getByText, getByRole } = render(
            <SideSheetRight title="Product Overview" isVisible={true}>
                <ProductOverview detail={testDetail} addToCart={addCartMock} />
            </SideSheetRight>
        );

        const button = getByText("ADD TO CART");

        fireEvent.click(button);
        expect(addCartMock).toHaveBeenCalledTimes(2);
    });
});

describe("<EnhancedCartForm />", () => {
    test("handleSubmit called with the correct values", async () => {
        const handleSubmit = jest.fn();

        const { getByRole, getByText } = render(
            <EnhancedCartForm handleSubmit={handleSubmit} availableQuantity={3} />
        );

        fireEvent.mouseDown(getByRole("button", { name: "1" }));
        expect(getByText("3")).toBeInTheDocument();
    });
});

// test("click me opens the drawer", async () => {
//     const { getByText, getByRole, queryByText } = render(
//         <SideSheetRight detail={testDetail} addCartFunction={addCartMock} />
//     );

//     //Open Drawer
//     const openDrawer = getByText("click me");
//     fireEvent.click(openDrawer);

//     await waitFor(() => {
//         expect(getByRole("close")).toBeInTheDocument();
//     });

//     // Confirm open drawer
//     expect(queryByText("Product Overview")).not.toBeNull();
// });

// test("close the drawer works", async () => {
//     const { getByText, getByRole } = render(
//         <SideSheetRight detail={testDetail} addCartFunction={addCartMock} />
//     );
//     //Open Drawer
//     const openDrawer = getByText("click me");
//     fireEvent.click(openDrawer);

//     await waitFor(() => {
//         expect(getByRole("close")).toBeInTheDocument();
//     });

//     const returnArrow = getByRole("close");
//     fireEvent.click(returnArrow);

//     await waitForElementToBeRemoved(() => getByText("Product Overview"));
// });

// test("confirm that the correct value has been logged", async () => {
//     const cartTest = jest.fn();

//     const { getByText, getByRole, getByTestId } = render(
//         <SideSheetRight detail={testDetail} addCartFunction={cartTest} />
//     );

//     // open the drawer
//     const openDrawer = getByText("click me");
//     fireEvent.click(openDrawer);
//     await waitFor(() => {
//         expect(getByText("Product Overview")).toBeInTheDocument();
//     });

//     // open the selector

//     const contentInput = getByTestId("content-input");
//     fireEvent.change(contentInput, {
//         target: { value: "1" },
//     });

//     const button = getByText("ADD TO CART");
//     fireEvent.click(button);

//     await waitFor(() => {
//         expect(cartTest).toHaveBeenCalledWith("1");
//     });
// });
