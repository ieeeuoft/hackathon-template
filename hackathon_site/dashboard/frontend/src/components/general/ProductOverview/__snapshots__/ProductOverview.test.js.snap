// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Overview render 1`] = `
<DocumentFragment>
  <div />
</DocumentFragment>
`;
